x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,526 @@
+            byte[] upperBytes = upper == null ? MAX_VALUE : Arrays.copyOfRange(upper.bytes, upper.offset, 16);
+
+            // point query on the 16byte prefix
+            Query pointPrefixQuery = BinaryPoint.newRangeQuery(name(), lowerBytes, upperBytes);
+
+            Query termQuery = new TermRangeQuery(
+                name(),
+                lowerTerm == null ? null : lower,
+                upperTerm == null ? null : upper,
+                includeLower,
+                includeUpper
+            );
+
+            return new BooleanQuery.Builder().add(new BooleanClause(pointPrefixQuery, Occur.MUST))
+                .add(new BooleanClause(termQuery, Occur.MUST))
+                .build();
+        }
+    }
+
+    private String nullValue;
+    private BooleanFieldMapper prereleaseSubField;
+    private NumberFieldMapper majorVersionSubField;
+    private NumberFieldMapper minorVersionSubField;
+    private NumberFieldMapper patchVersionSubField;
+
+    private VersionStringFieldMapper(
+        String simpleName,
+        FieldType fieldType,
+        MappedFieldType mappedFieldType,
+        String nullValue,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        BooleanFieldMapper preReleaseMapper,
+        NumberFieldMapper majorVersionMapper,
+        NumberFieldMapper minorVersionMapper,
+        NumberFieldMapper patchVersionMapper
+    ) {
+        super(simpleName, fieldType, mappedFieldType, multiFields, copyTo);
+        this.nullValue = nullValue;
+        this.prereleaseSubField = preReleaseMapper;
+        this.majorVersionSubField = majorVersionMapper;
+        this.minorVersionSubField = minorVersionMapper;
+        this.patchVersionSubField = patchVersionMapper;
+    }
+
+    @Override
+    public VersionStringFieldType fieldType() {
+        return (VersionStringFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected VersionStringFieldMapper clone() {
+        return (VersionStringFieldMapper) super.clone();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        String versionString;
+        if (context.externalValueSet()) {
+            versionString = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                versionString = nullValue;
+            } else {
+                versionString = parser.textOrNull();
+            }
+        }
+
+        if (versionString == null) {
+            return;
+        }
+
+        EncodedVersion encoding = encodeVersion(versionString);
+        BytesRef encodedVersion = encoding.bytesRef;
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
+            Field field = new Field(fieldType().name(), encodedVersion, fieldType);
```

Comment:
That seems redundant. If I understand correctly, the encoded version is stored in three structures:

- Fully encoded version in `Postings`

- The first 16 bytes of the encoded version in `Points`

- Fully encoded version in `SortedSetDocValues`

The main benefit of indexing `Points` is that you don't need the postings. We can use them as an approximation and rely on the doc values to validate matches in a range query. Prefix, wildcard, regex are more tricky to handle with `Points` so I guess that is why you also create an indexed `Field`. However, I think we should only pick one strategy. The current approach is too costly imo, we should control the indexing cost more carefully.